File: maths_library/python/forward_differentiation.py
class ForwardAutoDiff:
    def __init__(self, real, dual):
        self.real = real
        self.dual = dual

    def __add__(self, other):
        real = self.real + other.real
        dual = self.dual + other.dual
        return ForwardAutoDiff(real, dual)

    def __mul__(self, other):
        real = self.real * other.real
        dual = self.real * other.dual + self.dual * other.real
        return ForwardAutoDiff(real, dual)


def diff(f, x):
    return f(ForwardAutoDiff(x, 1)).dual


def f(x):
    return x * x + ForwardAutoDiff(6, 0) * x + ForwardAutoDiff(-1, 0)


if __name__ == "__main__":
    x = 3
    print(diff(f, x))  # expected: 2x + 6 = 12

File: maths_library/cpp/forward_differentiation.cpp
#include <iostream>

class ForwardAutoDiff
{
public:
    double real;
    double dual;

    ForwardAutoDiff(double real, double dual = 0.0)
        : real(real), dual(dual) {}

    ForwardAutoDiff operator+(const ForwardAutoDiff& other) const {
        return ForwardAutoDiff(real + other.real, dual + other.dual);
    }

    ForwardAutoDiff operator*(const ForwardAutoDiff& other) const {
        return ForwardAutoDiff(
            real * other.real,
            real * other.dual + dual * other.real
        );
    }
};

typedef ForwardAutoDiff (*DiffFunc)(ForwardAutoDiff);

double diff(DiffFunc f, double x) {
    return f(ForwardAutoDiff(x, 1.0)).dual;
}

ForwardAutoDiff f(ForwardAutoDiff x) {
    return x * x + ForwardAutoDiff(6) * x + ForwardAutoDiff(-1);
}

int main()
{
    double x = 3;
    std::cout << diff(f, x) << std::endl;  // 12
    return 0;
}